### PR TITLE
Options validation

### PR DESCRIPTION
**Pull request by defrag**

### CODE REVIEW

File: zio-cli/shared/src/main/scala/zio/cli/Options.scala [hunk 1]
```diff
@@ -59,14 +63,19 @@ object Options {
...
+    def optional: Options[Option[A]] = copy(optionType = Optional(optionType))
 
-    def map[B](f: A => B): Options[B] = copy(flagType = Map(flagType, (a: A) => Right(f(a))))
+    def map[B](f: A => B): Options[B] = copy(optionType = Map(optionType, (a: A) => Right(f(a))))
 
     def mapTry[B](f: A => B): Options[B] =
-      copy(flagType = Map(flagType, (a: A) => scala.util.Try(f(a)).toEither.left.map(_.getMessage)))
+      copy(optionType = Map(optionType, (a: A) => scala.util.Try(f(a)).toEither.left.map(_.getMessage)))
+
+    def validate(args: List[String], opts: ParserOptions): IO[List[HelpDoc.Block], (List[String], A)] =
```
Comment (jdegoes) [inline zio-cli/shared/src/main/scala/zio/cli/Options.scala:75]: I think this function here should be responsible for parsing the name / aliases, and should delegate only the parsing of the value itself to `Type#validate`. This will eliminate duplication and is also a better design decision: because the type should not know about the structure of the option as a whole, only its own structure.

So I'd consolidate the logic here and then delegate to the type validator in here.

Note that we decided on the following format for options:

 - `--foo bar` for options that take parameters
 - `-f` for options that do not take parameters (`Toggle`) and which have single-letter `name` or `aliases`. Note that we would like this to work like `wc -clfx`, i.e. if you have lots of toggles with single-letter names, then you can combine them all using only one dash.

File: zio-cli/shared/src/main/scala/zio/cli/Options.scala [hunk 2]
```diff
@@ -85,18 +94,68 @@ object Options {
...
+  sealed trait Type[+A] {
+    def validate(args: List[String], supportedOptions: Vector[String]): IO[List[HelpDoc.Block], (List[String], A)]
+  }
   object Type {
 
-    final case class Toggle(negationName: Option[String], ifPresent: Boolean) extends Type[Boolean]
-    final case class Map[A, B](value: Type[A], f: A => Either[String, B])     extends Type[B]
-    final case class Optional[A](value: Type[A])                              extends Type[Option[A]]
-    final case class Primitive[A](primType: PrimType[A])                      extends Type[A]
+    final case class Toggle(negationName: Option[String], ifPresent: Boolean) extends Type[Boolean] {
```
Comment (jdegoes) [inline zio-cli/shared/src/main/scala/zio/cli/Options.scala:110]: Let's say you have an option with `Toggle(Some("V"), true)`, where the primary name of the `Single` is "v".

Then the interpretation is as follows:

 - If the option is present, then it produces `true`, because `ifPresent = true`. 
 - If the option is absent, then it produces `false`, because `ifPresent = true`.
 - If the option is not present, but the negative name is present (`-V`), then it produces `false`, because `ifPresent` is true.

Does this make sense?

Basically `Toggle` is for binary settings, which have a default, and which maybe have a way to negate them.

Comment (defrag) [reply zio-cli/shared/src/main/scala/zio/cli/Options.scala:137]: @jdegoes we can't leave Toggle as type here, because it has more intricacies and thus those implementations can't be fulfilled. It has more custom options traversing like Optional type. If we leave it as type, then we can test for value not being present, since this fires only if value is present indeed. I don't like the idea of duplicating operators for Toggle, but its a bit problematic data type atm.

Comment (jdegoes) [reply zio-cli/shared/src/main/scala/zio/cli/Options.scala:138]: Why don't you leave that one as `Type`, and I can work on fixing it myself? i.e. focus on the other ones for now.

Comment (defrag) [reply zio-cli/shared/src/main/scala/zio/cli/Options.scala:138]: Sounds good :wink:

File: zio-cli/shared/src/test/scala/zio/cli/OptionsSpec.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    
+    testM("validate option and get remainder with different ordering") {
+      val r = f.validate(List("--bar", "baz", "--firstname", "John", "--lastname", "Doe"), ParserOptions.default)
+      assertM(r)(equalTo(List("--bar", "baz", "--lastname", "Doe") -> "John"))
+    },
+
+    testM("validate when no valid values are passed") {
+      val r = f.validate(List("--lastname", "Doe"), ParserOptions.default)
+      val expected =
+        List(p(error("No option found!. Was expecting option: --firstname.")))
```
Comment (jdegoes) [inline zio-cli/shared/src/test/scala/zio/cli/OptionsSpec.scala:38]: This test is a bit fragile. I think we should test only that it fails, not what the message is.

File: zio-cli/shared/src/test/scala/zio/cli/OptionsSpec.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      assertM(r.either)(isLeft(equalTo(expected)))
+    },
+    testM("validate when option is passed, but not a following value") {
+      val r        = f.validate(List("--firstname"), ParserOptions.default)
+      val expected = List(p(error("Value for option --firstname was not found!")))
```
Comment (jdegoes) [inline zio-cli/shared/src/test/scala/zio/cli/OptionsSpec.scala:43]: Same here.

**Comment by jdegoes** [comment]
This ready for another review?

**Comment by defrag** [comment]
Yes, ready for review in scope of what we talked about, just the Toggle is not there and we would have to decide if we need different type for an option that supports more continuous flags or we need to explode them at higher level.
